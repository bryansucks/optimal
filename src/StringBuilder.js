/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import Builder from './Builder';

function isString(value: *): boolean {
  return (typeof value === 'string' && value !== '');
}

export default class StringBuilder extends Builder<string> {
  allowEmpty: boolean = false;

  constructor(defaultValue: string = '') {
    super('string', defaultValue);

    // Check emptiness
    this.addCheck(this.checkNotEmpty);
  }

  contains(token: string, index: number = 0): this {
    if (__DEV__) {
      this.invariant(
        isString(token),
        'Contains requires a non-empty string.',
      );
    }

    return this.addCheck(this.checkContains, token, index);
  }

  checkContains(path: string, value: *, token: string, index: number = 0) {
    if (__DEV__) {
      this.invariant(
        (value.indexOf(token, index) >= 0),
        `String does not include "${token}".`,
        path,
      );
    }
  }

  match(pattern: RegExp): this {
    if (__DEV__) {
      this.invariant(
        (pattern instanceof RegExp),
        'Match requires a regular expression to match against.',
      );
    }

    return this.addCheck(this.checkMatch, pattern);
  }

  checkMatch(path: string, value: *, pattern: RegExp) {
    if (__DEV__) {
      this.invariant(
        value.match(pattern),
        `String does not match pattern "${pattern.source}".`,
        path,
      );
    }
  }

  empty(): this {
    this.allowEmpty = true;

    return this;
  }

  checkNotEmpty(path: string, value: *) {
    if (__DEV__) {
      if (!this.allowEmpty) {
        this.invariant(
          isString(value),
          'String cannot be empty.',
          path,
        );
      }
    }
  }

  oneOf(list: string[] = []): this {
    if (__DEV__) {
      this.invariant(
        (Array.isArray(list) && list.length > 0 && list.every(isString)),
        'One of requires a non-empty array of strings.',
      );
    }

    return this.addCheck(this.checkOneOf, list);
  }

  checkOneOf(path: string, value: *, list: string[] = []) {
    if (__DEV__) {
      this.invariant(
        (list.indexOf(value) >= 0),
        `String must be one of: ${list.join(', ')}`,
        path,
      );
    }
  }
}

export function string(defaultValue: string = ''): StringBuilder {
  return new StringBuilder(defaultValue);
}
