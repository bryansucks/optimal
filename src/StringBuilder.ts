/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Builder from './Builder';

function isString(value: any): value is string {
  return typeof value === 'string' && value !== '';
}

export default class StringBuilder<T extends string = string> extends Builder<T> {
  allowEmpty: boolean = false;

  constructor(defaultValue?: T) {
    super('string', defaultValue || ('' as T));

    // Not empty by default
    if (__DEV__) {
      this.addCheck(this.checkNotEmpty);
    }
  }

  contains(token: string, index: number = 0): this {
    if (__DEV__) {
      this.invariant(isString(token), 'Contains requires a non-empty string.');
    }

    return this.addCheck(this.checkContains, token, index);
  }

  checkContains(path: string, value: T, token: string, index: number = 0) {
    if (__DEV__) {
      this.invariant(value.indexOf(token, index) >= 0, `String does not include "${token}".`, path);
    }
  }

  match(pattern: RegExp): this {
    if (__DEV__) {
      this.invariant(
        pattern instanceof RegExp,
        'Match requires a regular expression to match against.',
      );
    }

    return this.addCheck(this.checkMatch, pattern);
  }

  checkMatch(path: string, value: T, pattern: RegExp) {
    if (__DEV__) {
      this.invariant(
        !!value.match(pattern),
        `String does not match pattern "${pattern.source}".`,
        path,
      );
    }
  }

  empty(): this {
    if (__DEV__) {
      this.allowEmpty = true;
    }

    return this;
  }

  checkNotEmpty(path: string, value: T) {
    if (__DEV__) {
      if (!this.allowEmpty) {
        this.invariant(isString(value), 'String cannot be empty.', path);
      }
    }
  }

  oneOf<U extends string>(list: U[]) /* refine */ {
    if (__DEV__) {
      this.invariant(
        Array.isArray(list) && list.length > 0 && list.every(item => isString(item)),
        'One of requires a non-empty array of strings.',
      );
    }

    this.addCheck(this.checkOneOf, list);

    return (this as any) as StringBuilder<U>;
  }

  checkOneOf(path: string, value: T, list: T[]) {
    if (__DEV__) {
      this.invariant(list.indexOf(value) >= 0, `String must be one of: ${list.join(', ')}`, path);
    }
  }
}

export function string<T extends string = string>(defaultValue?: T) /* infer */ {
  return new StringBuilder<T>(defaultValue);
}
