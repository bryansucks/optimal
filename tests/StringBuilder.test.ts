import StringBuilder, { string } from '../src/StringBuilder';

const camelCase = 'fooBarBaz1';
const kebabCase = 'foo-bar-baz2';
const pascalCase = 'FooBarBaz3';
const snakeCase = 'foo_bar_baz4';

describe('StringBuilder', () => {
  let builder: StringBuilder;

  beforeEach(() => {
    builder = string();
  });

  describe('constructor()', () => {
    it('sets default value', () => {
      expect(string('bar').defaultValue).toBe('bar');
    });
  });

  describe('runChecks()', () => {
    it('errors if a non-string value is used', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.runChecks('key', 123, { key: 123 });
      }).toThrowErrorMatchingSnapshot();
    });

    it('returns default value if value is undefined', () => {
      expect(string('abc').runChecks('key', undefined, { key: undefined })).toEqual('abc');
    });

    it('returns default value from factory if value is undefined', () => {
      expect(
        string(() => 'XYZ').runChecks('key', undefined, {
          key: undefined,
        }),
      ).toEqual('XYZ');
    });
  });

  describe('camelCase()', () => {
    it('errors if less than 2 characters', () => {
      expect(() => {
        builder.camelCase();
        builder.runChecks('key', 'a', { key: 'a' });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if starts with a number', () => {
      expect(() => {
        builder.camelCase();
        builder.runChecks('key', '1', { key: '1' });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if in kebab case', () => {
      expect(() => {
        builder.camelCase();
        builder.runChecks('key', kebabCase, { key: kebabCase });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if in pascal case', () => {
      expect(() => {
        builder.camelCase();
        builder.runChecks('key', pascalCase, { key: pascalCase });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if in snake case', () => {
      expect(() => {
        builder.camelCase();
        builder.runChecks('key', snakeCase, { key: snakeCase });
      }).toThrowErrorMatchingSnapshot();
    });

    it('passes if in camel case', () => {
      expect(() => {
        builder.camelCase();
        builder.runChecks('key', camelCase, { key: camelCase });
      }).not.toThrow();
    });
  });

  describe('contains()', () => {
    it('errors if token is not string', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.contains(123);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if token is an empty string', () => {
      expect(() => {
        builder.contains('');
      }).toThrowErrorMatchingSnapshot();
    });

    it('adds a checker', () => {
      builder.contains('oo');

      expect(builder.checks[1]).toEqual({
        callback: builder.checkContains,
        args: ['oo', 0],
      });
    });
  });

  describe('checkContains()', () => {
    it('errors if value does not contain token', () => {
      expect(() => {
        builder.checkContains('key', 'bar', 'oo');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if value matches default value and builder is required', () => {
      expect(() => {
        builder.required().checkContains('key', '', 'oo');
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if value matches default value and builder is optional', () => {
      expect(() => {
        builder.checkContains('key', '', 'oo');
      }).not.toThrow('Invalid field "key". String does not include "oo".');
    });

    it('doesnt error if value contains token', () => {
      expect(() => {
        builder.checkContains('key', 'foo', 'oo');
      }).not.toThrow('Invalid field "key". String does not include "oo".');
    });
  });

  describe('kebabCase()', () => {
    it('errors if less than 2 characters', () => {
      expect(() => {
        builder.kebabCase();
        builder.runChecks('key', 'a', { key: 'a' });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if starts with a number', () => {
      expect(() => {
        builder.kebabCase();
        builder.runChecks('key', '1', { key: '1' });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if in camel case', () => {
      expect(() => {
        builder.kebabCase();
        builder.runChecks('key', camelCase, { key: camelCase });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if in pascal case', () => {
      expect(() => {
        builder.kebabCase();
        builder.runChecks('key', pascalCase, { key: pascalCase });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if in snake case', () => {
      expect(() => {
        builder.kebabCase();
        builder.runChecks('key', snakeCase, { key: snakeCase });
      }).toThrowErrorMatchingSnapshot();
    });

    it('passes if in kebab case', () => {
      expect(() => {
        builder.kebabCase();
        builder.runChecks('key', kebabCase, { key: kebabCase });
      }).not.toThrow();
    });
  });

  describe('match()', () => {
    it('errors if pattern is not a regex', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.match(123);
      }).toThrowErrorMatchingSnapshot();
    });

    it('adds a checker', () => {
      builder.match(/oo/u);

      expect(builder.checks[1]).toEqual({
        callback: builder.checkMatch,
        args: [/oo/u, ''],
      });
    });
  });

  describe('checkMatch()', () => {
    it('errors if value does not match pattern', () => {
      expect(() => {
        builder.checkMatch('key', 'bar', /oo/u);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if value matches default value and builder is required', () => {
      expect(() => {
        builder.required().checkMatch('key', '', /oo/u);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if value matches default value and builder is optional', () => {
      expect(() => {
        builder.checkMatch('key', '', /oo/u);
      }).not.toThrow('Invalid field "key". String does not match pattern "oo".');
    });

    it('doesnt error if value matches pattern', () => {
      expect(() => {
        builder.checkMatch('key', 'foo', /oo/u);
      }).not.toThrow('Invalid field "key". String does not match pattern "oo".');
    });
  });

  describe('notEmpty()', () => {
    it('adds a checker', () => {
      builder.notEmpty();

      expect(builder.checks[1]).toEqual({
        callback: builder.checkNotEmpty,
        args: [],
      });
    });
  });

  describe('checkNotEmpty()', () => {
    it('errors if value is empty', () => {
      expect(() => {
        builder.checkNotEmpty('key', '');
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('oneOf()', () => {
    it('errors if not an array', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.oneOf(123);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if array is empty', () => {
      expect(() => {
        builder.oneOf([]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if array contains a non-string', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.oneOf(['foo', 123]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('adds a checker', () => {
      builder.oneOf(['foo', 'bar', 'baz']);

      expect(builder.checks[1]).toEqual({
        callback: builder.checkOneOf,
        args: [['foo', 'bar', 'baz']],
      });
    });
  });

  describe('checkOneOf()', () => {
    it('errors if value is not in the list', () => {
      expect(() => {
        builder.checkOneOf('key', 'qux', ['foo', 'bar', 'baz']);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if value contains token', () => {
      expect(() => {
        builder.checkOneOf('key', 'foo', ['foo', 'bar', 'baz']);
      }).not.toThrow('Invalid field "key". String must be one of: foo, bar, baz');
    });
  });

  describe('pascalCase()', () => {
    it('errors if less than 2 characters', () => {
      expect(() => {
        builder.pascalCase();
        builder.runChecks('key', 'A', { key: 'A' });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if starts with a number', () => {
      expect(() => {
        builder.pascalCase();
        builder.runChecks('key', '1', { key: '1' });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if in camel case', () => {
      expect(() => {
        builder.pascalCase();
        builder.runChecks('key', camelCase, { key: camelCase });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if in kebab case', () => {
      expect(() => {
        builder.pascalCase();
        builder.runChecks('key', kebabCase, { key: kebabCase });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if in snake case', () => {
      expect(() => {
        builder.pascalCase();
        builder.runChecks('key', snakeCase, { key: snakeCase });
      }).toThrowErrorMatchingSnapshot();
    });

    it('passes if in pascal case', () => {
      expect(() => {
        builder.pascalCase();
        builder.runChecks('key', pascalCase, { key: pascalCase });
      }).not.toThrow();
    });
  });

  describe('sizeOf()', () => {
    it('adds a checker', () => {
      builder.sizeOf(3);

      expect(builder.checks[1]).toEqual({
        callback: builder.checkSizeOf,
        args: [3],
      });
    });
  });

  describe('checkSizeOf()', () => {
    it('errors if length doesnt match', () => {
      expect(() => {
        builder.checkSizeOf('key', '', 3);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if length matches', () => {
      expect(() => {
        builder.checkSizeOf('key', 'abc', 3);
      }).not.toThrow();
    });
  });

  describe('snakeCase()', () => {
    it('errors if less than 2 characters', () => {
      expect(() => {
        builder.snakeCase();
        builder.runChecks('key', 'a', { key: 'a' });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if starts with a number', () => {
      expect(() => {
        builder.snakeCase();
        builder.runChecks('key', '1', { key: '1' });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if in camel case', () => {
      expect(() => {
        builder.snakeCase();
        builder.runChecks('key', camelCase, { key: camelCase });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if in kebab case', () => {
      expect(() => {
        builder.snakeCase();
        builder.runChecks('key', kebabCase, { key: kebabCase });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if in pascal case', () => {
      expect(() => {
        builder.snakeCase();
        builder.runChecks('key', pascalCase, { key: pascalCase });
      }).toThrowErrorMatchingSnapshot();
    });

    it('passes if in snake case', () => {
      expect(() => {
        builder.snakeCase();
        builder.runChecks('key', snakeCase, { key: snakeCase });
      }).not.toThrow();
    });
  });

  describe('typeAlias()', () => {
    it('returns the type name', () => {
      expect(string().typeAlias()).toBe('string');
    });
  });
});
