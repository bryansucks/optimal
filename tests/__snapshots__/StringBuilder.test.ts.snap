// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StringBuilder checkContains() errors if value does not contain token 1`] = `"Invalid field \\"key\\". String does not include \\"oo\\"."`;

exports[`StringBuilder checkContains() errors if value matches default value and builder is required 1`] = `"Invalid field \\"key\\". String does not include \\"oo\\"."`;

exports[`StringBuilder checkMatch() errors if value does not match pattern 1`] = `"Invalid field \\"key\\". String does not match pattern \\"oo\\"."`;

exports[`StringBuilder checkMatch() errors if value matches default value and builder is required 1`] = `"Invalid field \\"key\\". String does not match pattern \\"oo\\"."`;

exports[`StringBuilder checkNotEmpty() errors if value is empty 1`] = `"Invalid field \\"key\\". String cannot be empty."`;

exports[`StringBuilder checkOneOf() errors if value is not in the list 1`] = `"Invalid field \\"key\\". String must be one of: foo, bar, baz"`;

exports[`StringBuilder contains() errors if token is an empty string 1`] = `"Contains requires a non-empty token."`;

exports[`StringBuilder contains() errors if token is not string 1`] = `"Contains requires a non-empty token."`;

exports[`StringBuilder match() errors if pattern is not a regex 1`] = `"Match requires a regular expression to match against."`;

exports[`StringBuilder oneOf() errors if array contains a non-string 1`] = `"One of requires a non-empty array of strings."`;

exports[`StringBuilder oneOf() errors if array is empty 1`] = `"One of requires a non-empty array of strings."`;

exports[`StringBuilder oneOf() errors if not an array 1`] = `"One of requires a non-empty array of strings."`;

exports[`StringBuilder runChecks() errors if a non-string value is used 1`] = `"Invalid field \\"key\\". Must be a string."`;
